### ISSUE TITLE
Support service removals

### ISSUE THREAD

**Issue by kriswuollett** [windy1/zeroconf-rs#67]
For a use case connecting to services automatically and disconnecting when no longer available, it would be great if we could get an event or callback when a service was removed as seen with `AVAHI_BROWSER_REMOVE` and perhaps failing to resolve service after a Bonjour TTL timeout. I think it means the service is still available if the mDNS record is identical?

While I did look at the Bonjour code, and added a `Hash`/`PartialEq` impl for `TxtRecord` for deriving it for `ServiceDiscovery`, and added an generic event callback,  I didn't spend much time figuring out how to support a the user-provided event loop poll timeout along with needing something like a per-second timeout internally to check cached entries.

While just something like a `service_removed` callback would be added here, I'd assume it would align better with `zeroconf-tokio` if it was a `service_event` for `{added, removed, ...}` instead.

In the meantime I can probably just recreate a browser upon a relevant user event, or some timeout on my library's side.

**Comment by DebianArch64** [comment]
> For a use case connecting to services automatically and disconnecting when no longer available, it would be great if we could get an event or callback when a service was removed as seen with `AVAHI_BROWSER_REMOVE` and perhaps failing to resolve service after a Bonjour TTL timeout. I think it means the service is still available if the mDNS record is identical?
> 
> While I did look at the Bonjour code, and added a `Hash`/`PartialEq` impl for `TxtRecord` for deriving it for `ServiceDiscovery`, and added an generic event callback, I didn't spend much time figuring out how to support a the user-provided event loop poll timeout along with needing something like a per-second timeout internally to check cached entries.
> 
> While just something like a `service_removed` callback would be added here, I'd assume it would align better with `zeroconf-tokio` if it was a `service_event` for `{added, removed, ...}` instead.
> 
> In the meantime I can probably just recreate a browser upon a relevant user event, or some timeout on my library's side.

Yep, if the MDNS record is identical and has a TTL of 0 it means it's saying goodbye. The framework doesn't seem to use the TTL at all, would be a great addition in the meantime to expose it.

In many scenarios automatically connecting to the service and if it disconnects you are guaranteed it's gone is the best way. Sadly my scenario uses a VPN trick on iOS that prevents me from making connections to the various devices securely (only one virtual interface allowed).

https://github.com/windy1/zeroconf-rs/blob/7129d3e8b2fa4a0f2a557ce47b606b61c861a1a6/zeroconf/src/bonjour/browser.rs#L237

**Comment by inetic** [comment]
I also need this. Started implementing it [here](https://github.com/windy1/zeroconf-rs/compare/master...inetic:zeroconf-rs:service-remove). It's Avahi only right now and a work in progress. Will likely look into Bonjour next.

If anyone has any suggestions feel free to let me know.

**Comment by windy1** [comment]
This was merged
